### DISCUSSION TITLE
MetaPhysicL error in ThreadedElementLoopBase

### DISCUSSION THREAD

**Discussion by czadeh** [idaholab/moose#22033] [Q&A General] [answered]
Hello,

I added a new material property to an application I have been working on and now I get an error when I run an input file with the new MP. 

I have 2 AD variables and now 4 MPs. My input file without the new MP runs fine, but now I am getting this error:

*** ERROR ***
We caught a MetaPhysicL error in ThreadedElementLoopBase. This is very likely due to AD not having a sufficiently large derivative container size. Please run MOOSE configure with the '--with-derivative-size=<n>' option

Is this normal? Should I change the derivative size? If so, how?

Thanks,
Cameron

**Comment by GiudGiud** [comment, marked as answer]
Hello

~~What do you mean by MP?~~  material property right?

This error can happen with AD. You should do what the prompt says. The default is 53 for the container size, you may try like 100 at first. 
In the moose/ folder, run `./configure --with-derivative-size=100`

Guillaume

**Reply by czadeh** [reply]
Sorry, I meant material property. I will try that and let you know.

**Reply by czadeh** [reply]
Still getting that error even when I keep increasing the derivative size up to 2000. Perhaps it is an issue with the new files I added?

**Reply by GiudGiud** [reply]
weird. Are you recompiling after the configure?

**Reply by czadeh** [reply]
I recompiled my project I was working on before I tried again.

**Reply by GiudGiud** [reply]
ok good.
The error in Metaphysicl is not the error the prompt mentions then.

Please follow these instructions here to get us a backtrace
https://mooseframework.inl.gov/application_development/debugging.html

also let s consider any floating point operation you have on AD numbers. Do you have exotic tanh2 etc functions?

**Reply by czadeh** [reply]
Funny you mention that, I use the tanh function on AD Reals in my calculation.

**Reply by GiudGiud** [reply]
can you drop it for testing?
It could be that

**Reply by czadeh** [reply]
I dropped the tanh but still had the issue... but then I broke down my function line by line and I found out this:

- when log and tanh functions are removed it works fine
- if there are either log or tanh functions it gives that error

**Reply by GiudGiud** [reply]
can you add checks that you are within the range of validity of these functions?
and the range where the derivative is well defined as well. if the derivative is no longer defined then that could be trouble

log is only trouble near 0 I think?

**Reply by lindsayad** [reply]
What METHOD are you compiling with? (Default is opt)

**Reply by czadeh** [reply]
> can you add checks that you are within the range of validity of these functions? and the range where the derivative is well defined as well. if the derivative is no longer defined then that could be trouble
> 
> log is only trouble near 0 I think?

The domain of tanh(x) is all reals, so there should not be an input issue there.

For log(x), the input must be > 0. Based on my setup, the input here should be positive (at least on the first step with the initial condition). 

I did some more digging and discovered this:

- If I just have `_myMatProp[_qp] = tanh(_Temperature[_qp]);` it compiles and runs
- If I modify the above to make the argument of tanh what it actually needs to be, it also runs
- Same with log

In fact, every individual line of the function compiles and runs when I split them up, regardless of tanh or log functions. But when all the terms are together it has problems running. I tried isolating the problem from there but it gives the same error even when I just multiply or add most any 2 terms together (there are 4 terms that all run fine on their own).

> What METHOD are you compiling with? (Default is opt)

I was using the default, opt.

Should I create the debug executable and proceed from there?

**Reply by czadeh** [reply]
How is the derivative involved? I am using the void computeQpProperties() method which essentially just assigns the material property value after some user-defined arithmetic, right?

**Reply by lindsayad** [reply]
I would recommend following the instructions that @GiudGiud linked to: https://mooseframework.inl.gov/application_development/debugging.html That will eliminate the guesswork

**Reply by czadeh** [reply]
I followed those steps and the first thing that comes up is this (on time step 1):
```
Thread 1 "turtle-dbg" received signal SIGFPE, Arithmetic exception.
0x00007fffeb9093b4 in __feraiseexcept_invalid_divbyzero (
    __excepts=1) at ../sysdeps/x86/fpu/bits/fenv.h:135
135	../sysdeps/x86/fpu/bits/fenv.h: No such file or directory.
```
This is the back trace:
```
#0  0x00007fffeb9093b4 in __feraiseexcept_invalid_divbyzero (
    __excepts=1) at ../sysdeps/x86/fpu/bits/fenv.h:135
#1  __GI___feraiseexcept (__excepts=1)
    at ../sysdeps/x86/fpu/include/bits/fenv.h:46
#2  __log_compat (x=-603373.42361265107) at w_log_compat.c:42
#3  0x00007ffff7f081eb in std::log<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, unsigned long, MetaPhysicL::NWrapper<53ul> >, true> (in=...)
    at /home/cameron/mambaforge3/envs/moose/libmesh/include/metaphysicl/dualnumber.h:642
#4  0x00007ffff7f00bb4 in NonIsoCryst::computeQpProperties (
    this=0x555555b28b50)
    at /home/cameron/projects/turtle/src/materials/NonIsoCryst.C:54
#5  0x00007ffff53ffe3c in Material::computeProperties (
    this=0x555555b28b50)
    at /home/cameron/projects/moose/framework/src/materials/Material.C:132
#6  0x00007ffff5c74fff in MaterialData::reinit<std::vector<std::shared_ptr<MaterialBase>, std::allocator<std::shared_ptr<MaterialBase> > > > (this=0x555555a5b840, mats=...)
    at /home/cameron/projects/moose/framework/build/header_symlinks/MaterialData.h:454
#7  0x00007ffff5c2c6ad in FEProblemBase::reinitMaterials (
    this=0x555555a3df80, blk_id=0, tid=0, swap_stateful=true)
    at /home/cameron/projects/moose/framework/src/problems/FEProblemBase.C:3311
#8  0x00007ffff5621475 in NonlinearThread::onElement (
    this=0x7fffffffa3e0, elem=0x555555935f80)
    at /home/cameron/projects/moose/framework/src/loops/NonlinearThread.C:107
#9  0x00007ffff5672af5 in ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator() (this=0x7fffffffa3e0, range=..., 
    bypass_threading=false)
    at /home/cameron/projects/moose/framework/build/header_sym
```

It looks like a divide by zero error inside of my material property file NonIsoCryst.C. That is something I will definitely look into.

Because I have temperature in the denominator on some functions, I was interested to see its value.
I tried `p T` to get the temperature variable value, but I'm not sure how to interpret this output:

**Reply by GiudGiud** [reply]
If the log is evaluated at 0, that will do it. 

What you pasted at the bottom of your message (and I edited it up) was the rest of the backtrace (it sometimes gets printed slowly).
Maybe you can try again?

if the local variable in the code is called T, then `p T` should return the local value

**Reply by czadeh** [reply]
I set some bounds on the functions so there are no more divide by zero errors.

gdb helped me so much in figuring out the problems. I should have done that first! Thanks for your help.

**Reply by czadeh** [reply]
Now, I sometimes get a different error during runtime. Do you know what this one means?

When I back-trace in gdb it says no stack.

```
Stack frames: 20
0: libMesh::print_trace(std::ostream&)
1: moose::internal::mooseErrorRaw(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >)
2: void mooseError<char const (&) [219]>(char const (&) [219])
3: ThreadedElementLoopBase<libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> >::operator()(libMesh::StoredRange<libMesh::MeshBase::const_element_iterator, libMesh::Elem const*> const&, bool)
4: /home/cameron/projects/moose/framework/libmoose-dbg.so.0(+0xa7d74d) [0x7ffff524174d]
5: void AuxiliarySystem::computeElementalVarsHelper<AuxKernelTempl<double> >(MooseObjectWarehouse<AuxKernelTempl<double> > const&, std::vector<std::vector<MooseVariableFieldBase*, std::allocator<MooseVariableFieldBase*> >, std::allocator<std::vector<MooseVariableFieldBase*, std::allocator<MooseVariableFieldBase*> > > > const&)
6: AuxiliarySystem::computeElementalVars(MooseEnumItem)
7: AuxiliarySystem::compute(MooseEnumItem)
8: FEProblemBase::computeAuxiliaryKernels(MooseEnumItem const&)
9: FEProblemBase::execute(MooseEnumItem const&)
10: FixedPointSolve::solveStep(double&, double&, std::set<unsigned long, std::less<unsigned long>, std::allocator<unsigned long> > const&)
11: FixedPointSolve::solve()
12: TimeStepper::step()
13: Transient::takeStep(double)
14: Transient::execute()
15: MooseApp::executeExecutioner()
16: MooseApp::run()
17: main
18: __libc_start_main
19: /home/cameron/projects/turtle/turtle-dbg(+0x3079) [0x555555557079]
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
:
system msg for write_line failure : Bad file descriptor
[Thread 0x7fffea65a8c0 (LWP 68241) exited]
[Inferior 1 (process 68241) exited with code 01]
```

**Reply by GiudGiud** [reply]
you ll need to set a breakpoint on MPI_Abort if you want a backtrace

this is an error, so it exited properly with an error.
Doesnt it appear above?

**Reply by czadeh** [reply]
That error appears also when I run it in -opt mode, underneath the message:

```
*** ERROR ***
We caught a MetaPhysicL error in ThreadedElementLoopBase. This is very likely due to AD not having a sufficiently large derivative container size. Please run MOOSE configure with the '--with-derivative-size=<n>' option
```

When I use gdb and I set the breakpoint on MPI_Abort, back-trace just says no stack. So I can't figure out why exactly it's failing.

**Reply by GiudGiud** [reply]
Interesting.

So this error is asking for you to reconfigure and recompile with a larger AD size. 

That may or may not be the problem here (insufficient AD size), worth trying to exclude it

Guillaume

**Reply by czadeh** [reply]
I set the derivative size to 100 and it works now. So it wasn't the main problem originally but ended up being part of the solution.

If I increase the number of nodes it seems to require a greater derivative size.

**Reply by lindsayad** [reply]
In the future if you want to catch exceptions in gdb, you can do something like
```
b FEProblemBase::solve
run
```
and then when you hit that breakpoint
```
catch throw
run
```
You can't do `catch throw` at the beginning of the simulation because we rely on exception throwing in our Parser system
